File: go/policies.go
package workhorse

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"
)

// ErrInvalidPolicyResult reports a policy row that violates the public result contract.
var ErrInvalidPolicyResult = errors.New(invalidPolicyResultMessage)

// ErrInvalidPolicyOptions reports an unsupported synchronization option combination.
var ErrInvalidPolicyOptions = errors.New(invalidPolicyOptionsMessage)

// SyncPolicyOptions controls desired-state policy reconciliation.
type SyncPolicyOptions struct {
	Prune bool
}

// ConcurrencyPolicyDefinition is one queue's deployment-synchronized active-job budget.
type ConcurrencyPolicyDefinition struct {
	Queue           string `json:"queue"`
	MaxActive       int    `json:"maxActive"`
	MaxActivePerKey *int   `json:"maxActivePerKey,omitempty"`
}

// ConcurrencyPolicy is one persisted queue concurrency policy.
type ConcurrencyPolicy struct {
	Namespace       string
	Queue           string
	MaxActive       int
	MaxActivePerKey *int
	UpdatedAt       time.Time
}

// RateLimit is one continuously refilled PostgreSQL token bucket.
type RateLimit struct {
	Limit      int `json:"limit"`
	IntervalMS int `json:"intervalMs"`
	Burst      int `json:"burst"`
}

// RateLimitPolicyDefinition is one queue's deployment-synchronized start-rate budget.
type RateLimitPolicyDefinition struct {
	Queue  string     `json:"queue"`
	Rate   RateLimit  `json:"rate"`
	PerKey *RateLimit `json:"perKey,omitempty"`
}

// RateLimitPolicy is one persisted and normalized queue rate-limit policy.
type RateLimitPolicy struct {
	Namespace string
	Queue     string
	Rate      RateLimit
	PerKey    *RateLimit
	UpdatedAt time.Time
}

// SyncConcurrencyPolicies atomically reconciles one namespace of concurrency policies.
// Omitted definitions are removed unless options explicitly set Prune to false.
func (queue *Queue) SyncConcurrencyPolicies(
	ctx context.Context,
	namespace string,
	definitions []ConcurrencyPolicyDefinition,
	options ...SyncPolicyOptions,
) ([]ConcurrencyPolicy, error) {
	prune, err := policyPruneOption(options)
	if err != nil {
		return nil, err
	}
	if definitions == nil {
		definitions = []ConcurrencyPolicyDefinition{}
	}
	payload, err := json.Marshal(definitions)
	if err != nil {
		return nil, err
	}
	if err := AssertSchemaCompatible(ctx, queue.executor); err != nil {
		return nil, err
	}
	rows, err := queue.executor.Query(
		ctx,
		internalStatementRegistry[syncConcurrencyPoliciesStatementName],
		namespace,
		payload,
		prune,
	)
	if err != nil {
		return nil, err
	}
	return concurrencyPolicyRows(rows)
}

// ListConcurrencyPolicies returns persisted policies ordered by queue name.
// A nil or empty queue-name slice returns every policy.
func (queue *Queue) ListConcurrencyPolicies(
	ctx context.Context,
	queueNames []string,
) ([]ConcurrencyPolicy, error) {
	if queueNames == nil {
		queueNames = []string{}
	}
	rows, err := queue.executor.Query(
		ctx,
		internalStatementRegistry[listConcurrencyPoliciesStatementName],
		queueNames,
	)
	if err != nil {
		return nil, err
	}
	return concurrencyPolicyRows(rows)
}

// SyncRateLimitPolicies atomically reconciles one namespace of rate-limit policies.
// Omitted definitions are removed unless options explicitly set Prune to false.
func (queue *Queue) SyncRateLimitPolicies(
	ctx context.Context,
	namespace string,
	definitions []RateLimitPolicyDefinition,
	options ...SyncPolicyOptions,
) ([]RateLimitPolicy, error) {
	prune, err := policyPruneOption(options)
	if err != nil {
		return nil, err
	}
	if definitions == nil {
		definitions = []RateLimitPolicyDefinition{}
	}
	payload, err := json.Marshal(definitions)
	if err != nil {
		return nil, err
	}
	if err := AssertSchemaCompatible(ctx, queue.executor); err != nil {
		return nil, err
	}
	rows, err := queue.executor.Query(
		ctx,
		internalStatementRegistry[syncRateLimitPoliciesStatementName],
		namespace,
		payload,
		prune,
	)
	if err != nil {
		return nil, err
	}
	return rateLimitPolicyRows(rows)
}

// ListRateLimitPolicies returns persisted policies ordered by queue name.
// A nil or empty queue-name slice returns every policy.
func (queue *Queue) ListRateLimitPolicies(
	ctx context.Context,
	queueNames []string,
) ([]RateLimitPolicy, error) {
	if queueNames == nil {
		queueNames = []string{}
	}
	rows, err := queue.executor.Query(
		ctx,
		internalStatementRegistry[listRateLimitPoliciesStatementName],
		queueNames,
	)
	if err != nil {
		return nil, err
	}
	return rateLimitPolicyRows(rows)
}

func policyPruneOption(options []SyncPolicyOptions) (bool, error) {
	if len(options) > 1 {
		return false, fmt.Errorf(tooManySyncPolicyOptionsMessage, ErrInvalidPolicyOptions)
	}
	if len(options) == 1 {
		return options[0].Prune, nil
	}
	return true, nil
}

func concurrencyPolicyRows(rows []Row) ([]ConcurrencyPolicy, error) {
	policies := make([]ConcurrencyPolicy, len(rows))
	for index, row := range rows {
		namespace, namespaceOK := row[rowNamespaceField].(string)
		queueName, queueOK := row[rowQueueNameField].(string)
		maxActive, maxActiveOK := integer(row[rowMaxActiveField])
		maxActivePerKey, maxActivePerKeyOK := optionalInteger(row[rowMaxActivePerKeyField])
		updatedAt, updatedAtOK := row[rowUpdatedAtField].(time.Time)
		if !namespaceOK {
			return nil, invalidPolicyField(rowNamespaceField)
		}
		if !queueOK {
			return nil, invalidPolicyField(rowQueueNameField)
		}
		if !maxActiveOK {
			return nil, invalidPolicyField(rowMaxActiveField)
		}
		if !maxActivePerKeyOK {
			return nil, invalidPolicyField(rowMaxActivePerKeyField)
		}
		if !updatedAtOK {
			return nil, invalidPolicyField(rowUpdatedAtField)
		}
		policies[index] = ConcurrencyPolicy{
			Namespace: namespace, Queue: queueName, MaxActive: maxActive,
			MaxActivePerKey: maxActivePerKey, UpdatedAt: updatedAt,
		}
	}
	return policies, nil
}

func rateLimitPolicyRows(rows []Row) ([]RateLimitPolicy, error) {
	policies := make([]RateLimitPolicy, len(rows))
	for index, row := range rows {
		namespace, namespaceOK := row[rowNamespaceField].(string)
		queueName, queueOK := row[rowQueueNameField].(string)
		rate, rateOK := rateLimitFromRow(row, rowRateLimitField, rowRateIntervalMSField, rowRateBurstField)
		perKey, perKeyOK := optionalRateLimitFromRow(
			row,
			rowPerKeyLimitField,
			rowPerKeyIntervalMSField,
			rowPerKeyBurstField,
		)
		updatedAt, updatedAtOK := row[rowUpdatedAtField].(time.Time)
		if !namespaceOK {
			return nil, invalidPolicyField(rowNamespaceField)
		}
		if !queueOK {
			return nil, invalidPolicyField(rowQueueNameField)
		}
		if !rateOK {
			return nil, invalidPolicyField(rowRateLimitField)
		}
		if !perKeyOK {
			return nil, invalidPolicyField(rowPerKeyLimitField)
		}
		if !updatedAtOK {
			return nil, invalidPolicyField(rowUpdatedAtField)
		}
		policies[index] = RateLimitPolicy{
			Namespace: namespace, Queue: queueName, Rate: rate, PerKey: perKey, UpdatedAt: updatedAt,
		}
	}
	return policies, nil
}

func rateLimitFromRow(row Row, limitField, intervalField, burstField string) (RateLimit, bool) {
	limit, limitOK := integer(row[limitField])
	interval, intervalOK := integer(row[intervalField])
	burst, burstOK := integer(row[burstField])
	return RateLimit{Limit: limit, IntervalMS: interval, Burst: burst}, limitOK && intervalOK && burstOK
}

func optionalRateLimitFromRow(
	row Row,
	limitField string,
	intervalField string,
	burstField string,
) (*RateLimit, bool) {
	if row[limitField] == nil && row[intervalField] == nil && row[burstField] == nil {
		return nil, true
	}
	rate, ok := rateLimitFromRow(row, limitField, intervalField, burstField)
	return &rate, ok
}

func invalidPolicyField(field string) error {
	return fmt.Errorf(invalidPolicyFieldFormat, ErrInvalidPolicyResult, field)
}

File: rust/src/policies.rs
//! Deployment-synchronized concurrency, rate limit and budget policies.
use chrono::{DateTime, Utc};
use serde_json::{json, Value};
use tokio_postgres::Row;

/// A token bucket: `limit` starts per `interval_ms`, with up to `burst` saved.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct RateLimit {
    pub limit: i32,
    pub interval_ms: i32,
    pub burst: i32,
}

impl RateLimit {
    fn document(self) -> Value {
        json!({"limit": self.limit, "intervalMs": self.interval_ms, "burst": self.burst})
    }

    fn read(row: &Row, prefix: &str, limit: &str) -> Result<Option<Self>, tokio_postgres::Error> {
        let Some(limit) = row.try_get::<_, Option<i32>>(limit)? else {
            return Ok(None);
        };
        Ok(Some(Self {
            limit,
            interval_ms: row.try_get(format!("{prefix}interval_ms").as_str())?,
            burst: row.try_get(format!("{prefix}burst").as_str())?,
        }))
    }
}

/// Desired active-task limits for one queue.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ConcurrencyPolicyDefinition {
    pub queue: String,
    pub max_active: i32,
    pub max_active_per_key: Option<i32>,
}

/// A concurrency policy as PostgreSQL stores it.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ConcurrencyPolicy {
    pub namespace: String,
    pub queue: String,
    pub max_active: i32,
    pub max_active_per_key: Option<i32>,
    pub updated_at: DateTime<Utc>,
}

/// Desired start rate for one queue, and optionally for each concurrency key.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RateLimitPolicyDefinition {
    pub queue: String,
    pub rate: RateLimit,
    pub per_key: Option<RateLimit>,
}

/// A rate limit policy as PostgreSQL stores it.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RateLimitPolicy {
    pub namespace: String,
    pub queue: String,
    pub rate: RateLimit,
    pub per_key: Option<RateLimit>,
    pub updated_at: DateTime<Utc>,
}

/// A named limit shared by every task that selects it, across queues.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BudgetDefinition {
    pub name: String,
    pub max_active: Option<i32>,
    pub rate: Option<RateLimit>,
}

/// A budget as PostgreSQL stores it.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Budget {
    pub namespace: String,
    pub name: String,
    pub max_active: Option<i32>,
    pub rate: Option<RateLimit>,
    pub updated_at: DateTime<Utc>,
}

pub(crate) fn concurrency_document(definitions: &[ConcurrencyPolicyDefinition]) -> Value {
    definitions
        .iter()
        .map(|definition| {
            json!({
                "queue": definition.queue,
                "maxActive": definition.max_active,
                "maxActivePerKey": definition.max_active_per_key,
            })
        })
        .collect()
}

pub(crate) fn rate_limit_document(definitions: &[RateLimitPolicyDefinition]) -> Value {
    definitions
        .iter()
        .map(|definition| {
            json!({
                "queue": definition.queue,
                "rate": definition.rate.document(),
                "perKey": definition.per_key.map(RateLimit::document),
            })
        })
        .collect()
}

pub(crate) fn budget_document(definitions: &[BudgetDefinition]) -> Value {
    definitions
        .iter()
        .map(|definition| {
            json!({
                "name": definition.name,
                "maxActive": definition.max_active,
                "rate": definition.rate.map(RateLimit::document),
            })
        })
        .collect()
}

pub(crate) fn concurrency_policy(row: &Row) -> Result<ConcurrencyPolicy, tokio_postgres::Error> {
    Ok(ConcurrencyPolicy {
        namespace: row.try_get("namespace")?,
        queue: row.try_get("queue_name")?,
        max_active: row.try_get("max_active")?,
        max_active_per_key: row.try_get("max_active_per_key")?,
        updated_at: row.try_get("updated_at")?,
    })
}

pub(crate) fn rate_limit_policy(row: &Row) -> Result<RateLimitPolicy, tokio_postgres::Error> {
    Ok(RateLimitPolicy {
        namespace: row.try_get("namespace")?,
        queue: row.try_get("queue_name")?,
        rate: RateLimit {
            limit: row.try_get("rate_limit")?,
            interval_ms: row.try_get("rate_interval_ms")?,
            burst: row.try_get("rate_burst")?,
        },
        per_key: RateLimit::read(row, "per_key_", "per_key_limit")?,
        updated_at: row.try_get("updated_at")?,
    })
}

pub(crate) fn budget(row: &Row) -> Result<Budget, tokio_postgres::Error> {
    Ok(Budget {
        namespace: row.try_get("namespace")?,
        name: row.try_get("budget_name")?,
        max_active: row.try_get("max_active")?,
        rate: RateLimit::read(row, "rate_", "rate_limit")?,
        updated_at: row.try_get("updated_at")?,
    })
}
